aten/src/ATen/native/Blas.cpp

```diff
@@ -137,6 +140,110 @@ Tensor dot(const Tensor &self, const Tensor &other){
   });
 }
 
+// this is to port LDA_COND from the TH codebase
+// #define LDA_COND(M, N, LDA) ((N) == 1 || (LDA) >= THMax(1, (M)))
+// lda_cond function in this file is not used
+// because it requires `lda > std::max<int64_t>(1L, m)`
+// instead of `>=` as defined in the TH implementation
+constexpr inline bool lda_cond_addr(int64_t m, int64_t n, int64_t lda) {
+  return n == 1 || lda >= std::max<int64_t>(1L, m);
+}
+
+template<typename scalar_t>
+Tensor &addr_impl_cpu(Tensor &result, const Tensor &self,
+                    const Tensor& vec1, const Tensor& vec2,
+                    scalar_t beta, scalar_t alpha) {
+  if (&result != &self) {
+    at::native::resize_as_(result, self);
+    at::native::copy_(result, self);
+  }
+  if (beta == 0.0) {
+    at::native::zero_(result);
+  } else if (beta != 1.0) {
+    at::native::mul_(result, beta);
+  }
+
+  if (result.stride(0) == 1 &&
+    lda_cond_addr(vec1.size(0), vec2.size(0), result.stride(1))) {
+    addr_impl<scalar_t>(
+      vec1.size(0), vec2.size(0),
+      alpha, vec1.data_ptr<scalar_t>(), vec1.stride(0),
+      vec2.data_ptr<scalar_t>(), vec2.stride(0),
+      result.data_ptr<scalar_t>(), result.stride(1)
+    );
+  } else if (result.stride(1) == 1 &&
+    lda_cond_addr(vec2.size(0), vec1.size(0), result.stride(0))) {
+    addr_impl<scalar_t>(
+      vec2.size(0), vec1.size(0),
+      alpha, vec2.data_ptr<scalar_t>(), vec2.stride(0),
+      vec1.data_ptr<scalar_t>(), vec1.stride(0),
+      result.data_ptr<scalar_t>(), result.stride(0)
+    );
+  } else {
+    Tensor cr = result.clone();
+    addr_impl<scalar_t>(
+      vec2.size(0), vec1.size(0),
+      alpha, vec2.data_ptr<scalar_t>(), vec2.stride(0),
+      vec1.data_ptr<scalar_t>(), vec1.stride(0),
+      cr.data_ptr<scalar_t>(), cr.stride(0)
+    );
+    result.set_(cr);
+  }
+  return result;
+}
+
+Tensor& addr_out_cpu(Tensor &result, const Tensor& self,
+                      const Tensor& vec1, const Tensor& vec2,
+                      Scalar beta, Scalar alpha) {
+  TORCH_CHECK(vec1.dim() == 1 && vec2.dim() == 1,
+              "vec1 and vec2 should be 1-dimensional vectors. Got dimensions ",
+              vec1.dim(), " and ", vec2.dim());
+
+  Tensor self_;
+  if (&result != &self) {
+    std::tie(self_) = expand_size(self, {vec1.size(0), vec2.size(0)}, "addr");
+  } else {
+    self_ = self;
+  }
+
+  TORCH_CHECK(result.device() == self_.device() &&
```

Comment:
If this was written as a composite op we would also get all this error checking for free ;)